examples/exex/op-proposer/src/main.rs

```diff
@@ -61,56 +51,141 @@ pub struct L2Output {
+        &std::fs::read_to_string("config.json").expect("Could not read config file"),
+    )
+    .unwrap();
 
-async fn op_proposer_exex<Node: FullNodeComponents>(
-    mut ctx: ExExContext<Node>,
-) -> eyre::Result<()> {
-    while let Some(notification) = ctx.notifications.recv().await {
-        match &notification {
-            ExExNotification::ChainCommitted { new } => {
-                info!(committed_chain = ?new.range(), "Received commit");
-            }
-            ExExNotification::ChainReorged { old, new } => {
-                info!(from_chain = ?old.range(), to_chain = ?new.range(), "Received reorg");
-            }
-            ExExNotification::ChainReverted { old } => {
-                info!(reverted_chain = ?old.range(), "Received revert");
-            }
-        };
-    }
-    Ok(())
+    let provider = ProviderBuilder::new()
+        .with_recommended_fillers()
+        .signer(EthereumSigner::from(config.proposer_private_key.parse::<LocalWallet>().unwrap()))
+        .on_http(config.l1_rpc.parse().unwrap());
+    Ok(OpProposer::new(
+        provider,
+        config.l2_output_oracle,
+        config.l2_to_l1_message_passer,
+        config.submission_interval,
+    )
+    .spawn(ctx, connection)?)
 }
 
-struct OpProposer {
-    //TODO: persistent storage configuration
+struct OpProposer<T, N, P>
+where
+    T: Transport + Clone,
+    N: Network,
+    P: Provider<T, N>,
+{
+    provider: Arc<P>,
+    l2_output_oracle: Address,
+    l2_to_l1_message_passer: Address,
+    submission_interval: u64,
+    _network: std::marker::PhantomData<N>,
+    _transport: std::marker::PhantomData<T>,
 }
 
-impl OpProposer {
-    fn new() -> Self {
-        Self {}
+impl<T: Transport + Clone, N: Network, P: Provider<T, N>> OpProposer<T, N, P> {
+    fn new(
+        provider: P,
+        l2_output_oracle: Address,
+        l2_to_l1_message_passer: Address,
+        submission_interval: u64,
+    ) -> Self {
+        Self {
+            provider: Arc::new(provider),
+            l2_output_oracle,
+            l2_to_l1_message_passer,
+            submission_interval,
+            _network: std::marker::PhantomData,
+            _transport: std::marker::PhantomData,
+        }
     }
 
     fn spawn<Node: FullNodeComponents>(
         &self,
         mut ctx: ExExContext<Node>,
+        _connection: Connection,
     ) -> eyre::Result<impl Future<Output = eyre::Result<()>>> {
-        //TODO: initialization logic
-
+        let l2_oo = L2OutputOracle::new(self.l2_output_oracle, self.provider.clone());
+        let l2_provider = ctx.provider().clone();
         let fut = async move {
             while let Some(notification) = ctx.notifications.recv().await {
                 match &notification {
                     ExExNotification::ChainCommitted { new } => {
+                        let current_block = ctx.head.number;
+                        let next_block = l2_oo.latestBlockNumber().call().await?._0.to::<u64>();
```

Comment:
We can cache this outside of the loop so we dont have to call `latestBlockNumber` every iteration.